### PR TITLE
Modify ensemble to better reflect composing model response pattern

### PR DESCRIPTION
**Pull request by GuanLuo**
Previously ensemble will detect "decoupled" field and send responses that may not reflect the composing model's response pattern, i.e.:
if the composing model sends response and response complete flag separately,
```
TRITONBACKEND_ResponseSend(response, 0, nullptr /* success */);
TRITONBACKEND_ResponseFactorySendFlags(factory, TRITONSERVER_RESPONSE_COMPLETE_FINAL);
```
ensemble may send one response with the flag set or follow the same pattern, depending on whether "decoupled" is set in model config.

With this change the behavior will be more determined. And it reduces special handling in ensemble w.r.t. "decoupled" field in model config, which is meant to be used in server frontend (HTTP / gRPC).

### CODE REVIEW

File: src/ensemble_scheduler.cc [hunk 1]
```diff
@@ -1032,18 +1041,28 @@ EnsembleContext::FinishEnsemble(std::unique_ptr<InferenceResponse>&& response)
...
+      InferenceResponse::Send(std::move(response), flags);
+      response_sent_ = true;
+    } else if (flags != 0) {
+      // check if any response is sent, if no, decoupled must be specified
+      if ((!info_->is_decoupled_) && (!response_sent_)) {
+        ensemble_status_ = Status(
+            Status::Code::INVALID_ARG,
+            "in ensemble '" + info_->ensemble_name_ + "', " +
+                request_tracker_->Request()->LogRequest() +
+                "unexpected deadlock, at least one output is not set while no "
```
Comment (dyastremsky) [inline src/ensemble_scheduler.cc:1057]: Is there a way to detect a deadlock in decoupled models?

File: src/ensemble_scheduler.cc [hunk 2]
```diff
@@ -1032,18 +1041,28 @@ EnsembleContext::FinishEnsemble(std::unique_ptr<InferenceResponse>&& response)
   }
 
   if (ensemble_status_.IsOk()) {
-    if (info_->is_decoupled_) {
-      if (response != nullptr) {
-        InferenceResponse::Send(std::move(response), 0 /* flags */);
-      }
-      if (inflight_step_counter_ != 0) {
-        return ensemble_status_;
+    auto flags = (inflight_step_counter_ == 0)
```
Comment (oandreeva-nv) [inline src/ensemble_scheduler.cc:1044]: I am wondering of this `flags` is actually needed. We only use it on L1048. Should we also use it on L1064?

File: src/ensemble_scheduler.cc [hunk 3]
```diff
@@ -1100,14 +1121,8 @@ EnsembleContext::CheckAndSetEnsembleOutput(
...
   if (!ready) {
-    if (info_->is_decoupled_) {
-      return Status::Success;
-    }
-    return Status(
-        Status::Code::INVALID_ARG,
-        lrequest->LogRequest() +
-            "unexpected deadlock, at least one output is not set while no more "
-            "ensemble steps can be made");
+    // Do not create response if the actual response is not ready.
```
Comment (oandreeva-nv) [inline src/ensemble_scheduler.cc:1124]: So this was partially related to original issue?  We were creating a response, when it was not ready?

File: src/ensemble_scheduler.cc [hunk 4]
```diff
@@ -774,11 +781,13 @@ EnsembleContext::PrepareSteps(
         ensemble_status_ = GetNextSteps(updated_tensors, ready_steps);
       }
 
-      // Check and send ensemble response
-      if ((!ensemble_status_.IsOk()) || (inflight_step_counter_ == 0) ||
-          info_->is_decoupled_) {
+      // Reaching the end of processing completed step, check if the ensemble
+      // should send response / error.
+      if (!ensemble_status_.IsOk()) {
```
Comment (oandreeva-nv) [inline src/ensemble_scheduler.cc:786]: I see, so the previous logic was checking `ensemble_status_`, `inflight_step_counter_` and `info_->is_decoupled_ ` here. The new logic introduced a more precise behavior based on  `inflight_step_counter_` and `info_->is_decoupled_  values in combination with `response_sent_`.

File: src/ensemble_scheduler.cc [hunk 5]
```diff
@@ -340,6 +350,10 @@ class EnsembleContext {
   // Objects related to the ensemble infer request
   Status ensemble_status_;
   RequestTracker* request_tracker_;
+  // Use in conjunction with 'is_decoupled_' in EnsembleInfo to
+  // better distinguish ensemble ending behavior (see documentation in
+  // FinishEnsemble for detail).
```
Comment (oandreeva-nv) [inline src/ensemble_scheduler.cc:355]: My only complain here is that I was expecting to see **documentation**. I think `annotation` would be better for current case

File: src/ensemble_scheduler.cc (hunk 4 above)
Comment (GuanLuo) [reply src/ensemble_scheduler.cc:786]: Yes, previously the conditional will "trim" the number of possible ensemble responses, which causes the ensemble response not being filled properly.

**Review by oandreeva-nv — DISMISSED** [review]
LGTM

**Review by tanmayv25 — DISMISSED** [review]

**Review by oandreeva-nv — APPROVED** [review]

**Review by oandreeva-nv — APPROVED** [review]

**Pull request merged**